Settings: Refactor Post Formats to use `react-query`

tyxla opened the pull request:
#### Changes proposed in this Pull Request

Considering that post formats a very straightforward example of server state on the client side in Calypso, and the simplicity of the use case (read-only), this makes it a great example of how much boilerplate we can remove by using `react-query` instead of redux.

This PR refactors the post formats Redux state to use `react-query` and ports the one single usage of this state tree.

#### Testing instructions

* Make sure your WP.com site is using a theme that supports one or more custom post formats. You can find such a theme here: https://wordpress.com/themes/filter/post-formats/:site where `:site` is your WP.com site.
* Go to `/settings/writing/:site`
* Verify the post formats in the "Default post format" are loading correctly.
* Pick a site with a theme without any custom post formats. 
* Verify the field correctly displays just the "Standard" option and there are no errors in the console.

tyxla committed:
Settings: Refactor Post Formats to react-query

flootr commented on client/data/post-formats/use-post-formats-query.js:
```diff
@@ -0,0 +1,16 @@
+
+const usePostFormatsQuery = ( siteId ) =>
+	useQuery( [ 'postFormats', siteId ], () =>
+		wpcom.undocumented().site( siteId ).postFormatsList()
```
I think we started using `wpcom.req.get` directly so we can eventually move away from `wpcom.undocumented`.

tyxla replied:
Sure thing! Done in fa174d5bc515b11810f2d9d1a09f91731dfb8ddf. I've also removed the unused method in a0aad533a2820183e474c9feb75bdcf60b4e0c67.

flootr commented on client/my-sites/site-settings/composing/default-post-format.jsx:
```diff
@@ -42,14 +39,12 @@ const DefaultPostFormat = ( {
-						);
-					} ) }
+				{ data &&
+					map( data.formats, ( label, slug ) => (
```
Can we avoid using `_.map` in this case? I know it's preexisting to this PR but may be a good side-refactor.

tyxla replied:
I don't mind refactoring that as it's the only Lodash usage here. Removed in 36d2ed21465999c528175be8d95fa819d06f0d63.

flootr commented on client/my-sites/site-settings/composing/default-post-format.jsx:
```diff
@@ -69,11 +64,4 @@ DefaultPostFormat.propTypes = {
-		postFormats: getPostFormats( state, siteId ),
-	};
-} )( localize( DefaultPostFormat ) );
+export default localize( DefaultPostFormat );
```
As this is now a functional component what about using `useTranslate` instead?

tyxla replied:
👍  Done in 976158ac1266a37c058c203e938cde637c3a3e19.

flootr commented on client/my-sites/site-settings/composing/default-post-format.jsx:
```diff
@@ -1,37 +1,34 @@
-	siteId,
 	translate,
 } ) => {
+	const siteId = useSelector( ( state ) => getSelectedSiteId( state ) );
```
nit: could this be point-free `useSelector( getSelectedSiteId )` or would it have unintentional side effects?

tyxla replied:
Sure 👍  Done in fa174d5bc515b11810f2d9d1a09f91731dfb8ddf

flootr reviewed (commented):
I ❤️ this. Just a few comments. Let me know what you think.

tyxla committed:
Use wpcom.req.get directly

tyxla committed:
Remove UndocumentedSite.prototype.postFormatsList

tyxla committed:
Simplify useSelector usage

tyxla committed:
Refactor away from _.map()

tyxla committed:
localize() -> useTranslate()

tyxla commented:
> I ❤️ this. Just a few comments. Let me know what you think.

Thanks for the great review, @flootr! I've addressed all the feedback, could you take another look?

flootr reviewed (approved):
This works great 👍🏻 awesome refactor!

There's two things I want to mention though (I don't consider any of them as blocking):

- It seems like post formats used to be persisted (with schema validation), that's not the case with ReactQuery anymore (we haven't implemented this yet). However, I'm not sure if that's really needed for this kind of data. Looping in @jsnajdr as well as we should discuss on how to move on there.
- I'm not sure about the level of "freshness" we need for post formats. Right now data is refetched on focus change and its stale time is very low (that's the default). Post formats only change when the active theme is switched, correct? We could increase `staleTime` and/or disable `refetchOnWindowFocus` but I'm not sure about what works best and I guess this is something we should to iterate on in the future.

tyxla commented:
> It seems like post formats used to be persisted (with schema validation), that's not the case with ReactQuery anymore (we haven't implemented this yet). However, I'm not sure if that's really needed for this kind of data. Looping in @jsnajdr as well as we should discuss on how to move on there.

I don't think it's needed, TBH. I don't see much value in it, given how simple the use case is.

> I'm not sure about the level of "freshness" we need for post formats. Right now data is refetched on focus change and its stale time is very low (that's the default). Post formats only change when the active theme is switched, correct? We could increase staleTime and/or disable refetchOnWindowFocus but I'm not sure about what works best and I guess this is something we should to iterate on in the future.

I've been thinking about that and decided to keep the defaults, because that way if the user changes the theme and goes back to that window, they will be refetched. However, I'm happy to disable those if you think that default freshness is unnecessary and only adds extra request overhead.

That being said, I'll refrain from shipping this for now until we have a consensus here - let me know what you think @flootr @jsnajdr.

flootr commented:
@tyxla from my side this is ready to go, my comments were meant as general notes.

tyxla commented:
I'll go ahead and ship this then. Happy to follow up if you folks have suggestions.

Thanks again, Markus 🙌
